dev/UndockedRegFreeWinRT/typeresolution.cpp

```diff
@@ -428,40 +428,69 @@ namespace UndockedRegFreeWinRT
         _COM_Outptr_opt_result_maybenull_ IMetaDataImport2** ppMetaDataImport,
         _Out_opt_ mdTypeDef* pmdTypeDef)
     {
-        HRESULT hr = S_OK;
-        UINT32 dwPackagesCount = 0;
-        UINT32 dwBufferLength = 0;
-
-        const UINT32 filter = PACKAGE_FILTER_HEAD | PACKAGE_FILTER_DIRECT | PACKAGE_FILTER_IS_IN_RELATED_SET;
-        hr = HRESULT_FROM_WIN32(GetCurrentPackageInfo(filter, &dwBufferLength, nullptr, &dwPackagesCount));
-        // Only find the type if the it is an unpackaged app. Packaged apps can have their exe on their package graph,
-        // which will allow type resolution against adjacent WinMDs.
-        if (hr == HRESULT_FROM_WIN32(APPMODEL_ERROR_NO_PACKAGE))
+        // Walk the package graph looking for the requested metadata
+        const uint32_t filter{ PACKAGE_FILTER_HEAD | PACKAGE_FILTER_DIRECT | PACKAGE_FILTER_IS_IN_RELATED_SET | PACKAGE_FILTER_DYNAMIC | PACKAGE_FILTER_STATIC };
+        uint32_t bufferLength{};
+        uint32_t packagesCount{};
+        bool processHasStaticPackageGraph{};
+        HRESULT hr{ HRESULT_FROM_WIN32(GetCurrentPackageInfo(filter, &bufferLength, nullptr, &packagesCount)) };
+        if (hr == HRESULT_FROM_WIN32(ERROR_INSUFFICIENT_BUFFER))
```

Comment:
Since this code follows an early return pattern, it might be cleaner to finish the usage of hr very locally. It gets reused sporadically, which requires a bit more careful checking across the function.
{
 HRESULT hr{ HRESULT_FROM_WIN32(GetCurrentPackageInfo(filter, &bufferLength, nullptr, &packagesCount)) };
 if (hr != HRESULT_FROM_WIN32(ERROR_INSUFFICIENT_BUFFER))
 {
  return hr; // Also I think there's a will macro to still log / track the failure in this case
 }
}
// do the thing.

Reply:
Also, is there any chance of this succeeding with null buffer (e.g. can package info length be zero?)

Reply:
> Since this code follows an early return pattern, it might be cleaner to finish the usage of hr very locally

NP. `const HRESULT hr` and `hrFindBlahBlah` multiple/specific names are common practices to avoid reuse complexities.

Reply:
> return hr; // Also I think there's a will macro to still log / track the failure in this case

That would emit WIL messages even for lookups that work and successfully does-not-find what you're looking for. So we avoid false positive messages / noise but merely `return`.

But I changed the called `FindTypeInDirectoryWithNormalization` to use WIL checks (instead of silent `if (SUCCEEDED(hr))`) so if something does go _wrong_ we'll get some information. Probably should go over the whole URFW and update it to use WIL, but that's a bit out of scope here

Reply:
> is there any chance of this succeeding with null buffer (e.g. can package info length be zero?)

If `packageCount>0` then no, bufferLength is `>0`

No change of buffer overrun